### PR TITLE
feat: add lazy loading of tables

### PR DESCRIPTION
**Pull request by PengLiVectra**
# Description
Add lazy loading of tables. When preforming streaming operations we don't need any version of the table loaded. Large tables are slow to load, so we see a huge performance boost by avoiding CPU time spent loading the table metadata.

# Related Issue(s)
Partly with https://github.com/delta-io/delta-rs/issues/1361

# Testing
Adding Unit tests

# Breaking Change
Not a breaking change.

### CODE REVIEW

File: python/src/lib.rs [hunk 1]
```diff
@@ -85,6 +85,28 @@ struct RawDeltaTableMetaData {
...
+    if let Some(buf_size) = log_buffer_size {
+        builder = builder
+            .with_log_buffer_size(buf_size)
+            .map_err(PythonError::from)?;
+    }
+    Ok(builder)
+}
+
 #[pymethods]
 impl RawDeltaTable {
```
Comment (ion-elgreco) [inline python/src/lib.rs:111]: I think this makes sense to add, just two comments, I think we can simplify it a lot. We can just add the load_lazy argument as input for the `new` method, something like this: 

```rust
   fn new(
        table_uri: &str,
        version: Option<i64>,
        storage_options: Option<HashMap<String, String>>,
        without_files: bool,
        log_buffer_size: Option<usize>,
        load_lazy: bool,
    ) -> PyResult<Self> {
        let options = storage_options.clone().unwrap_or_default();
        let mut builder = deltalake::DeltaTableBuilder::from_uri(table_uri);
        if let Some(storage_options) = storage_options {
            builder = builder.with_storage_options(storage_options)
        }
        if without_files {
            builder = builder.without_files()
        }
        if let Some(buf_size) = log_buffer_size {
            builder = builder
                .with_log_buffer_size(buf_size)
                .map_err(PythonError::from)?;
        }
        if let Some(version) = version {
            builder = builder.with_version(version)
        }
        let table = if !load_lazy {
            rt()?.block_on(builder.load()).map_err(PythonError::from)?
        } else {
            builder.build().map_err(PythonError::from)?
        };
        Ok(RawDeltaTable {
            _table: table,
            _config: FsConfig {
                root_url: table_uri.into(),
                options,
            },
        })
```
and in python we just do an if else to set the metadata:
```python
        if load_lazy:
            self._metadata = None
        else:
            self._metadata = Metadata(self._table)
```

**Comment by roeap** [comment]
@PengLiVectra - could you explain the use case a bit? 

For typical scenarios, without protocol and metadata information there is not much we can (legally) do with the table.

To avoid loading the state into memory we can already pass options to omit files and tombstones. Currently we are also working on making that more lazy - i.e. just getting protocol and metadata asap, but avoid loading anything else.

**Comment by PengLiVectra** [comment]
> @PengLiVectra - could you explain the use case a bit?
> 
> For typical scenarios, without protocol and metadata information there is not much we can (legally) do with the table.
> 
> To avoid loading the state into memory we can already pass options to omit files and tombstones. Currently we are also working on making that more lazy - i.e. just getting protocol and metadata asap, but avoid loading anything else.

Yes, when using in streaming operations, sometimes the task has no data, it takes too much time to load large tables. We did a simple test with `load_lazy`:
# cat test.py
`import deltalake
x = deltalake.DeltaTable(s3path, lazy_load=True)
print(x.peek_next_commit(4690))`

If `load_lazy = True`, result is `python test.py  0.89s user 0.39s system 84% cpu 1.526 total`
If `load_lazy = False`, result is `python test.py  97.53s user 3.74s system 87% cpu 1:56.23 total`

**Comment by ion-elgreco** [comment]
@PengLiVectra if the task has no data, and you can check this why not simply load the table state if there is data, if not you skip the step.

**Comment by roeap** [comment]
I still don't quite understand the use case :) 

If your task receives no data, then there should be no need to lad the table at all, right? But when does hte need to peek the next table version arise?

delta-rs was initially build to support large scale log ingestion into delta via kafka streams (this is also why the kafka-delta-ingest project exists)

To handle the sometimes very large state that gets generated there are two options that can be passed to the table `require_files` and `require_tombstones` which will cause the internals to not load the file actions (we don't want to do a scan) and the tombstones (we also don't want to vacuum).

for any other meaningful interaction with the table, we always need the protocol and metadata actions. for the above scenario it would also need to be an `appendOnly` table.

So from the litle I know about your use case, I would suggest: 
1. configure table as append only.
2. check you data if you need to write to the table at all
3. if you need to write  data, load the table with `require_files` and `require_tombstones` set to false?

Would that meet your needs?

**Comment by PengLiVectra** [comment]
> 3\. require_files

Sorry for the misleading @ion-elgreco and @roeap , we didn't know if there is no data prior, we need to load the table and check if there is a new commit. So basically we need to save the time in the step of loading table.

**Comment by ion-elgreco** [comment]
@PengLiVectra also in that case, wouldn't initializing the table without files give the same result combined with get_latest_version or peek next commit

**Comment by PengLiVectra** [comment]
> @PengLiVectra if the task has no data, and you can check this why not simply load the table state if there is data, if not you skip the step.



> @PengLiVectra also in that case, wouldn't initializing the table without files give the same result combined with get_latest_version or peek next commit



> @PengLiVectra also in that case, wouldn't initializing the table without files give the same result combined with get_latest_version or peek next commit

Thanks, I'm still not very clear. Could you give an example about it in Python?

**Comment by PengLiVectra** [comment]
> I still don't quite understand the use case :)
> 
> If your task receives no data, then there should be no need to lad the table at all, right? But when does hte need to peek the next table version arise?
> 
> delta-rs was initially build to support large scale log ingestion into delta via kafka streams (this is also why the kafka-delta-ingest project exists)
> 
> To handle the sometimes very large state that gets generated there are two options that can be passed to the table `require_files` and `require_tombstones` which will cause the internals to not load the file actions (we don't want to do a scan) and the tombstones (we also don't want to vacuum).
> 
> for any other meaningful interaction with the table, we always need the protocol and metadata actions. for the above scenario it would also need to be an `appendOnly` table.
> 
> So from the litle I know about your use case, I would suggest:
> 
> 1. configure table as append only.
> 2. check you data if you need to write to the table at all
> 3. if you need to write  data, load the table with `require_files` and `require_tombstones` set to false?
> 
> Would that meet your needs?

`require_files` and `require_tombstones` are in `build` function, not in `load` function
https://github.com/delta-io/delta-rs/blob/main/crates/deltalake-core/src/table/builder.rs#L280-L300.

It seems only `load` functions is exposed in python (https://github.com/delta-io/delta-rs/blob/main/python/src/lib.rs#L116).

Is there a way to use this in Python?

**Comment by ion-elgreco** [comment]
@PengLiVectra that's likely because it doesn't load any state and also doesn't validates the file path unlike the normal load would do

@roeap please correct me if I'm wrong

**Comment by roeap** [comment]
@PengLiVectra - sorry for the large delay.

We are currently working towards release 0.17 which will include many breaking changes. One large piece of work I am currently addressing is to optimize the whole log handling and be as minimal as possible in getting to a useful state of the delta table. We are also leaking a lot of internals that we cannot commit to long term.

Why am I telling you this? :) 

> TL;DR; would it be OK to wait with a way forward a bit until some internal changes have landed? 

Your use case makes a lot of sense and we certainly have the opportunity to provide an implementation that streaming use cases can benefit from - much like you proposed.

The lazy-load in this case is more on the object store level, since if we do not have any metadata and without it we cannot do anything meaningful specific to delta. In fact the fact that the metadata and protool are currently optional is a major nuicance, that permeates the entire codebase ... IIRC, you are most interested if a new version of the table exists, right?

In that case we will also have facilities to inspect the state of the log from a storage perspective. and we may expose some of this for use cases like yours.

**Comment by PengLiVectra** [comment]
> @PengLiVectra - sorry for the large delay.
> 
> We are currently working towards release 0.17 which will include many breaking changes. One large piece of work I am currently addressing is to optimize the whole log handling and be as minimal as possible in getting to a useful state of the delta table. We are also leaking a lot of internals that we cannot commit to long term.
> 
> Why am I telling you this? :)
> 
> > TL;DR; would it be OK to wait with a way forward a bit until some internal changes have landed?
> 
> Your use case makes a lot of sense and we certainly have the opportunity to provide an implementation that streaming use cases can benefit from - much like you proposed.
> 
> The lazy-load in this case is more on the object store level, since if we do not have any metadata and without it we cannot do anything meaningful specific to delta. In fact the fact that the metadata and protool are currently optional is a major nuicance, that permeates the entire codebase ... IIRC, you are most interested if a new version of the table exists, right?
> 
> In that case we will also have facilities to inspect the state of the log from a storage perspective. and we may expose some of this for use cases like yours.

Thanks, @roeap. Do you have a timeline for release 0.17?

**Comment by ion-elgreco** [comment]
@PengLiVectra now that we already using the new log replay in python as well, I believe this can be closed. Does it meet your needs?
